### PR TITLE
[MachineLearning.Service] Add Service APIs for Machine Learning 

### PR DESCRIPTION
**Pull request by niley7464**
[MachineLearning.Inference] Add MlInformation and MlInformationList class
[MachineLearning.Service] Add Service APIs for Machine Learning

### CODE REVIEW

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/AssemblyAttr.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Tizen.MachineLearning.Service, " + PublicKey.Value)]
+
+internal static class PublicKey
```
Comment (niley7464) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/AssemblyAttr.cs:22]: This file makes Tizen.MachineLearning.Service to access internal function of Tizen.MachineLearning.Inference

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ */
+
+using System;
+
+namespace Tizen.MachineLearning.Inference
+{
+    /// <summary>
+    /// The MlInformation class manages information based on key-value.
+    /// </summary>
+    /// <since_tizen> 9 </since_tizen>
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: plz check version (latest version of c# api would be 12)

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+            switch(_type)
+            {
+                case InfoType.Option:
+                    IntPtr valuePtr = Interop.Util.StringToIntPtr(value);
+                    ret = Interop.Util.SetOptionValue(_handle, key, valuePtr, null);
+                    break;
+                case InfoType.Information:
+                    ret = NNStreamerError.NotSupported;
+                    Log.Error(NNStreamer.TAG, "InfoType Iniformation does not support set value");
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs:198]: typo: Iniformation

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        /// <summary>
+        /// Gets a TensorsData value of key in mlInformation instance.
+        /// </summary>
+        /// <param name="key">The key to be set.</param>
+        /// <returns>The TensorsData value of the key.</returns>
+        /// <feature>http://tizen.org/feature/machine_learning.inference</feature>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <since_tizen> 12 </since_tizen>
+        public TensorsData GetTensorsData(string key, TensorsInfo info)
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Consider adding new API (internal or public?) to get tensors-info by using 'ml_tensors_data_get_info()' from data handle.
Then we can get the tensors-data instance without tensors-info instance.
```
public TensorsData GetTensorsData(string key)
```

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+using System;
+
+namespace Tizen.MachineLearning.Inference
+{
+    /// <summary>
+    /// The MlInformation class manages information based on key-value.
+    /// </summary>
+    /// <since_tizen> 12 </since_tizen>
+    public class MlInformation
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: For MlInformation and MlInformationList classes, do we implement IDisposal if?
We can add internal flag to separate instance creation.
```
// fromList means MlInfomation from MlInformationList, no need to release native handle.
internal MlInformation(IntPtr handle, bool fromList) {
}
```

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            NNStreamer.CheckException(ret, "Failed to create information handle");
+
+            _type = InfoType.Option;
+        }
+
+        /// <summary>
+        /// Creates a MlInformation instance from Native handle.
+        /// </summary>
+        /// <since_tizen> 12 </since_tizen>
+        internal MlInformation(IntPtr handle) {
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: recommend adding param type for future work :)
```
internal MlInformationFromNativeHandle(IntPtr handle, InfoType type, bool fromList)
```

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 4 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Oh, I really needed this API while adding TCs.
I think this makes easier to handle tensor-data.
I wil add this API and updates some TCs.
Thank you :)

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 3 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs:198]: Updated!
Thank you

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 2 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Updated!
Thank you

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+                    EventReceived?.Invoke(this, new ServiceReceivedEventArgs(info));
+                }
+            };
+
+            MLServiceError ret = Interop.Service.SetEventCb(_handle, _eventCallback, IntPtr.Zero);
+            MLService.CheckException(ret, "Failed to register event callback");
+        }
+
+        /// <summary>
+        /// Destroys the MlInformation resource.
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Destroys machine learning service instance.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+        public event EventHandler<ServiceReceivedEventArgs> EventReceived;
+        private Interop.Service.EventCallback _eventCallback;
+
+        public enum EventType
+        {
+            NewData = 1,
+        }
+
+        /// <summary>
+        /// Creates a Service instance.
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Creates new machine learning service from configuration file.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+        {
+            MLServiceError ret = Interop.Service.Stop(_handle);
+            MLService.CheckException(ret, "Failed to stop the service because of internal error");
+        }
+
+        /// <summary>
+        /// Gets the TensorsInfo of the service instance.
+        /// </summary>
+        /// <param name="IsInput">The boolean value for choosing input or output to get.</param>
+        /// <param name="name">The name of node in the pipeline. You can set NULL if serivce is constructed from model configuration.</param>
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: The name of node in the pipeline. You can set empty string if machine learning service is constructed from model configuration.

Ditto for Request() method.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+        /// <feature>http://tizen.org/feature/machine_learning.service</feature>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
+        /// <since_tizen> 12 </since_tizen>
+        public TensorsInfo GetTensorsInformation(bool IsInput, string name = "")
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: param name rule: IsInput -> isInput

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
+        }
+
+        /// <summary>
+        /// Sets the information of the service instance.
+        /// </summary>
+        /// <param name="name">The name to be set.</param>
+        /// <param name="value">The value to be set.</param>
+        /// <feature>http://tizen.org/feature/machine_learning.service</feature>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs:160]: should add invalid-param exception.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+        }
+
+        /// <summary>
+        /// Gets the information of the service instance.
+        /// </summary>
+        /// <param name="name">The name to get the value.</param>
+        /// <returns>The string value of the name.</returns>
+        /// <feature>http://tizen.org/feature/machine_learning.service</feature>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs:182]: also need to add invalid-param exception.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+
+        /// <summary>
+        /// Adds an input data to process the model in service instance.
+        /// </summary>
+        /// <param name="data">The data to be processed.</param>
+        /// <param name="name">The name of node in the pipeline. You can set NULL if serivce is constructed from model configuration.</param>
+        /// <feature>http://tizen.org/feature/machine_learning.service</feature>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when failed to stop the service.</exception>
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs:204]: need invalid-param exception.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs [hunk 14]
```diff
@@ -0,0 +1,696 @@
...
+                    Log.Error(MLService.TAG, "Failed to destroy the service handle");
+
+                _handle = IntPtr.Zero;
+            }
+
+            _disposed = true;
+        }
+
+        /// <summary>
+        /// The Query class provides service based on MlInformation.
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: The Query class provides machine learning service for edge connection.

let's consider using key-word edge or connection to remote device.

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 5 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Updated!
I never release MlInformation by mistake 😅 Thank you :)

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 6 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Updated :)

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs (hunk 7 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Updated!

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs (hunk 8 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Updated!

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs (hunk 9 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Updated!

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs (hunk 10 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Updated!

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs (hunk 11 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs:160]: I think `ArgumentException` is invalid-param exception.
There was no InvalidParameter exception in the other examples either.

File: src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs (hunk 14 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Service/Tizen.MachineLearning.Service/Service.cs]: Updated! Thank you

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformationList.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+            // release unmanaged objects
+            if (_handle != IntPtr.Zero)
+            {
+                NNStreamerError ret = NNStreamerError.InvalidParameter;
+
+                if (_handle != IntPtr.Zero)
+                {
+                    ret = Interop.Util.DestroyInformationList(_handle);
+                    NNStreamer.CheckException(ret, "Failed to destroy the information list");
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformationList.cs]: IMO unnecessary line.

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformationList.cs [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+                }
+
+                NNStreamer.CheckException(ret, "Failed to destroy the information");
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformationList.cs]: plz fix exception message "Failed to destroy the information list"

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+            if (string.IsNullOrEmpty(key))
+                throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidParameter, "The property key is invalid");
+
+            IntPtr value = IntPtr.Zero;
+            NNStreamerError ret = NNStreamerError.InvalidParameter;
+
+            switch(_type)
+            {
+                case InfoType.Option:
+                    ret = Interop.Util.GetOptionValue(_handle, key, out value);
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Q: need conversion 'key' to intptr?

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+            if (_disposed)
+                return;
+
+            if (disposing)
+            {
+                // release managed object
+            }
+
+            // release unmanaged objects
+            if (_handle != IntPtr.Zero)
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs:104]: This line is unnecessary condition, duplicated w/line 108.
Need to remove one of these.

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+                    break;
+            }
+
+            NNStreamer.CheckException(ret, "Failed to get information handle from key");
+
+            return value;
+        }
+
+        /// <summary>
+        /// Gets a string value of key in mlInformation instance.
```
Comment (jaeyun-jung) [inline src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Plz fix class name for all method descriptions,
`mlInformation instance -> MlInformation instance`

Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: Updated it!
Thank you

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 18 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs:104]: Oh,, my bad 😭
Removed it
Thank you

File: src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs (hunk 17 above)
Comment (niley7464) [reply src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/MlInformation.cs]: The `key` value is only string type now, but as we discuss earlier I updated it to `IntPtr`
Thank you :)

**Pull request closed without merging**